Changed Content-Disposition header to download Contact Photo with correct extension

jacobneplokh opened the pull request:
Now prompts to download as ContactPhoto.png cross-browser

Signed-off-by: Jacob Neplokh <me@jacobneplokh.com>

--
Fixes #20143 
@jancborchardt @skjnldsv 

The `Content-Disposition` header now also has the file name (set to `ContactPhoto.png` — if this needs to be changed, let me know) to make sure it is downloaded with the correct file extension, and across browsers.

rullzer commented on apps/dav/lib/CardDAV/ImageExportPlugin.php:
```diff
@@ -103,7 +103,7 @@ public function httpGet(RequestInterface $request, ResponseInterface $response)
 			$file = $this->cache->get($addressbook->getResourceId(), $node->getName(), $size, $node);
 			$response->setHeader('Content-Type', $file->getMimeType());
-			$response->setHeader('Content-Disposition', 'attachment');
+			$response->setHeader('Content-Disposition', 'attachment; filename="ContactPhoto.png"');
```
I'm don't fully know the code around it. But are we sure it is always a png?

skjnldsv replied:
No, we need to use the mimetype

skjnldsv replied:
I would use `$node->getName()` for the filename and maybe https://github.com/nextcloud/server/blob/cb057829f72c70e819f456edfadbb29d72dba832/apps/dav/lib/CardDAV/PhotoCache.php#L45-L50 to get the extension based on `$file->getMimeType()` ?

jneplokh replied:
Ah true, good point. I'll go look into it.

jneplokh replied:
Hey @rullzer and @skjnldsv ,

I made some changes using the suggestions. As this is a "good first issue," I chose this to get some practice and learn some more :)

So let me know if there is something to change/more to do. Also, I noticed the PHP Code-Standards check failed, but am having a little trouble understanding where specifically the error is. 

Thank you!

jneplokh commented:
Hey @rullzer and @skjnldsv ,

I made some changes using the suggestions. As this is a "good first issue," I chose this to get some practice and learn some more :)

So let me know if there is something to change/more to do. Also, I noticed the PHP Code-Standards check failed, but am having a little trouble understanding where specifically the error is. 

Thank you!

skjnldsv commented on apps/dav/lib/CardDAV/ImageExportPlugin.php:
```diff
@@ -103,7 +104,7 @@ public function httpGet(RequestInterface $request, ResponseInterface $response)
 			$file = $this->cache->get($addressbook->getResourceId(), $node->getName(), $size, $node);
 			$response->setHeader('Content-Type', $file->getMimeType());
-			$response->setHeader('Content-Disposition', 'attachment');
+			$response->setHeader('Content-Disposition', 'attachment; filename="'.$node->getName().'.'.PhotoCache::ALLOWED_CONTENT_TYPES[$file->getMimeType()].'"');
```
```suggestion
			$fileName = $node->getName() . '.' . PhotoCache::ALLOWED_CONTENT_TYPES[$file->getMimeType()];
			$response->setHeader('Content-Disposition', "attachment; filename='$fileName'");
```
Maybe cleaner? :)

jneplokh replied:
Good idea, just committed those changes to use the `$fileName` variable

skjnldsv replied:
> Also, what should I do about the failing php-cs-check and build?

Run `composer i` and `composer run cs:fix`

skjnldsv commented:
@jneplokh please also rebase and squash your commits into one :)

skjnldsv commented:
There is a failure on the carddav integration test  `build/integration/features/carddav.feature:53`
`Expected value 'attachment' for header 'Content-Disposition', got 'attachment; filename=bjoern.vcf.jpg' (Exception)`

jneplokh commented:
Ok, I fixed the error on the carddav integration test by changing the expected header. To clarify, those 3 other failures (`nodb`, `postgres11-php7.2`, and `integration-auth`) are false positives correct?

skjnldsv commented:
> 1) OCA\DAV\Tests\unit\CardDAV\ImageExportPluginTest::testCard with data set #1 (null, true)
> Undefined index: imgtype
> 
> /drone/src/apps/dav/lib/CardDAV/ImageExportPlugin.php:106
> /drone/src/apps/dav/tests/unit/CardDAV/ImageExportPluginTest.php:198
> 
> 2) OCA\DAV\Tests\unit\CardDAV\ImageExportPluginTest::testCard with data set #3 (32, true)
> Undefined index: imgtype
> 
> /drone/src/apps/dav/lib/CardDAV/ImageExportPlugin.php:106
> /drone/src/apps/dav/tests/unit/CardDAV/ImageExportPluginTest.php:198


This seems related though

jneplokh commented:
Hmm, true. Maybe something with not getting the full mimetype but just the extension. I'll see what is going on there.

jneplokh commented:
Hey,

Sorry for the late comment here, got busy with something for a few weeks. I went over the PHPUnit docs, but I am honestly lost on what to change in the test. I got some great help from the Contacts channel, but am still confused on how to make the test work with the changes. 

What should the next course of action be? If it is closing the PR, I understand.

Thank you for all the help, though, during this PR!

skjnldsv commented:
> What should the next course of action be? If it is closing the PR, I understand.

No no! :)
Next step is to fix the step and get this in :wink: 

Look at the line of the test https://github.com/nextcloud/server/blob/2a054e6c04e0a40421510eb889cbf59f153c5177/apps/dav/tests/unit/CardDAV/ImageExportPluginTest.php#L177-L182

Since you changed the headers, the unit tests needs to reflect those changes too :)
You can run your test locally by running `phpunit --bootstrap apps/dav/tests/unit/bootstrap.php apps/dav/tests/unit/CardDAV/ImageExportPluginTest.php`

jneplokh commented:
> No no! :)
Next step is to fix the step and get this in 😉

Awesome!

> Since you changed the headers, the unit tests needs to reflect those changes too

Ok yeah, I thought this was happening but was not exactly sure *where* to change it. Thanks for the testing command too.

I'll keep working on it :)

jneplokh commented:
> Since you changed the headers, the unit tests needs to reflect those changes too

I only changed the `Content-Disposition` header, though, and not the `Content-Type`. Or is it because I changed the file name?

If so, do I need to change both places were `imgtype` is referenced?

jacobneplokh committed:
Change Content-Disposition Header

- Make ALLOWED_CONTENT_TYPES public in order to use
- Add $fileName variable which uses "$node->getName()" to get the proper file name and "$file->getMimeType()" along with the ALLOWED_CONTENT_TYPES array in PhotoCache.php to get the proper file extension
- Make "$fileName" the name of the file in the Content-Disposition header when downloading a Contact's photo
- Add filename to the CardDAV integration image export test header
- Change headers in ImageExportPluginTest to reflect changes

Signed-off-by: Jacob Neplokh <me@jacobneplokh.com>

skjnldsv commented:
> Also, are those two failures false positives?

Looks like it :)

jneplokh commented:
> Looks like it :)

Awesome! So is it effectively ready to be merged?

skjnldsv commented:
Thank you @jneplokh !!

jneplokh commented:
No problem! Definitely learned, too.

And thanks for the help during it! :)
